### PR TITLE
[Return-types #12.3] JAX-JIT interface higher-order derivatives

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
Allows the JAX-JIT interface to work with higher-order derivatives using the new return type system. Works with gradient transforms.

**Description of the Change:**
Changes the JAX-JIT interface accordingly.

**Benefits:**

Hessians via `jax.hessian` or by applying `jax.jacrev`/`jax.fwdrev` (or the mix of thereof) twice work on jitted QNodes.

**Possible Drawbacks:**
If the user mistakenly specifies `max_diff=1` and requests the Hessian via `jax.hessian`, then a non-informative JAX error is raised:
image

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/interfaces/jax_jit_tuple.py [hunk 1]
```diff
@@ -214,15 +211,42 @@ def execute_wrapper_jvp(primals, tangents):
...
+                        execute_fn,
+                        gradient_fn,
+                        gradient_kwargs,
+                        _n=_n + 1,
+                        max_diff=max_diff,
+                    )
+                    jacs = res_processing_fn(jacs)
+                    all_jacs.append(jacs)
+
+                if len(all_jacs) == 1 and len(tangents[0]) > 1:
```
Comment (rmoyard) [inline pennylane/interfaces/jax_jit_tuple.py:242]: Does it represent the case of executing one tape with multiple parameters?

File: tests/returntypes/jax/test_jax_jit_qnode_new.py [hunk 2]
```diff
@@ -2056,3 +2040,267 @@ def circuit(a):
 
         assert isinstance(jac[1], jax.numpy.ndarray)
         assert jac[1].shape == (4, 2)
+
+
+hessian_fn = [
```
Comment (rmoyard) [inline tests/returntypes/jax/test_jax_jit_qnode_new.py:2045]: Thank you for parametrizing that it is a very good idea!

File: tests/returntypes/jax/test_jax_jit_qnode_new.py [hunk 3]
```diff
@@ -2056,3 +2040,267 @@ def circuit(a):
...
+
+        par_0 = jax.numpy.array(0.1, dtype=jax.numpy.float64)
+        par_1 = jax.numpy.array(0.2, dtype=jax.numpy.float64)
+
+        @qnode(dev, interface="jax", diff_method=diff_method, max_diff=2)
+        def circuit(x, y):
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliX(1)), qml.probs(wires=[0, 1])
```
Comment (rmoyard) [inline tests/returntypes/jax/test_jax_jit_qnode_new.py:2190]: Great achievement, amazing!!!

**Review by rmoyard — COMMENTED** [review]
It is an amazing achievement :1st_place_medal: Only a question, also could we add an example to return type doc?

File: pennylane/interfaces/jax_jit_tuple.py (hunk 1 above)
Comment (rmoyard) [reply pennylane/interfaces/jax_jit_tuple.py:242]: Also now I see that it is not covered, so maybe it is not necessary?

File: pennylane/interfaces/jax_jit_tuple.py [hunk 4]
```diff
@@ -214,15 +211,42 @@ def execute_wrapper_jvp(primals, tangents):
...
+                    all_jacs = all_jacs[0]
+
+                jvps = _compute_jvps(all_jacs, tangents[0], multi_measurements)
```
Comment (PhilipVinc) [inline pennylane/interfaces/jax_jit_tuple.py:245]: I think ideally you'd want to return the jacobian itself from the callback, not the `jvp`.

This is because if you only compute the jacobian in the callback, but perform the jacobioan-vector product in "jax" code, jax will trace that product and he will be able to transpose it (it's just a product, so it's linear).

Being able to transpose it means that jax will be able to compute the reverse differentiation (vjp) automatically, simply by swapping the `jacobian-vector` of this function with `vector-jacobian` products.

I believe if you did that, you'd be able to have a single rule for both forward and backward instead of 2 different rules.

In short, I'd suggest to return `all_jacs` from the callback and run `_compute_jvps` in jax-traced code.

File: tests/returntypes/jax/test_jax_jit_qnode_new.py [hunk 5]
```diff
@@ -2056,3 +2040,267 @@ def circuit(a):
...
+
+        hess = jax.jit(hessian(circuit))(params)
+
+        assert isinstance(hess, tuple)
+        assert len(hess) == 2
+        assert isinstance(hess[0], jax.numpy.ndarray)
+        assert hess[0].shape == (
+            2,
+            2,
+        )
```
Comment (eddddddy) [inline tests/returntypes/jax/test_jax_jit_qnode_new.py:2303]:
```suggestion
        assert hess[0].shape == (2, 2)
```

File: tests/returntypes/jax/test_jax_jit_qnode_new.py [hunk 6]
```diff
@@ -2056,3 +2040,267 @@ def circuit(a):
...
+        """The hessian of single a measurement with multiple params return a tuple of arrays."""
+        dev = qml.device(dev_name, wires=2)
+
+        if diff_method == "adjoint":
+            pytest.skip("Test does not supports adjoint because second order diff.")
+
+        par_0 = jax.numpy.array(0.1)
+        par_1 = jax.numpy.array(0.2)
+
+        @qnode(dev, interface="jax", diff_method=diff_method, max_diff=2)
```
Comment (eddddddy) [inline tests/returntypes/jax/test_jax_jit_qnode_new.py:2068]: Is there a difference between using `interface="jax-jit"` directly in the QNode and using `interface="jax"` in the QNode but decorating the hessian function with `jax.jit`? In the changelog we have an example with the former, but these tests do the latter.

**Review by eddddddy — APPROVED** [review]
Looks good after @rmoyard and @PhilipVinc's comments are addressed 👍

File: pennylane/interfaces/jax_jit_tuple.py (hunk 4 above)
Comment (antalszava) [reply pennylane/interfaces/jax_jit_tuple.py:245]: Hi @PhilipVinc thanks for the comment :slightly_smiling_face: 

If I understand the suggestion correctly, that should already be the case:
* `all_jacs` is the result of the function evaluated using the callback;
* `_compute_jvps` that computs JVPs doesn't perform callbacks anymore;
* both `jax.jacrev` and `jax.jacfwd` should work now.

The reason why there's a "backward" and "forward" mode implementation is specific to the adjoint differentiation method to computing quantum gradients. The adjoint differentiation technique is suited to simultaneously evaluate the quantum function and compute the Jacobian on QNode evaluation. This is not always desired, however, hence PennyLane offers two modes.

Comment (PhilipVinc) [reply pennylane/interfaces/jax_jit_tuple.py:245]: > all_jacs is the result of the function evaluated using the callback;

Ops sorry, I missed that. Then it looks great!

File: pennylane/interfaces/jax_jit_tuple.py (hunk 1 above)
Comment (antalszava) [reply pennylane/interfaces/jax_jit_tuple.py:242]: Yes, I think originally, I added this for the one tape, multiple parameters case. However, it seems there's no need for it. Removing it now, thanks for the catch.

File: tests/returntypes/jax/test_jax_jit_qnode_new.py (hunk 6 above)
Comment (antalszava) [reply tests/returntypes/jax/test_jax_jit_qnode_new.py:2068]: [With `interface="jax"` PennyLane attempts to determine](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/interfaces/jax.py#L30) if the JAX-Python or the JAX-JIT interface should be used. The implementation works in a way, that `interface="jax"` and decorating a function with `jax.jit` will resolve to JAX-JIT.

When specifying `interface="jax-python"`/`interface="jax-jit"`, we definitely use the JAX-Python/JAX-JIT interface.

This is not a purely functional solution and is discouraged by JAX devs. Nonetheless, it brings user convenience to our users.

**Comment by antalszava** [comment]
Thank you for the comments! This is up for another review. (cc @rmoyard)

**Review by rmoyard — APPROVED** [review]
Thank you it's amazing :100:
